src/core/ext/gcp/metadata_query.cc

```diff
@@ -0,0 +1,127 @@
+
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+
+#include <grpc/grpc.h>
+#include <grpc/grpc_security.h>
+#include <grpc/support/log.h>
+
+#include "src/core/lib/debug/trace.h"
+#include "src/core/lib/gprpp/ref_counted_ptr.h"
+#include "src/core/lib/gprpp/status_helper.h"
+#include "src/core/lib/gprpp/time.h"
+#include "src/core/lib/security/credentials/credentials.h"
+#include "src/core/lib/uri/uri_parser.h"
+
+namespace grpc_core {
+
+TraceFlag grpc_metadata_query_trace(false, "metadata_query");
+
+constexpr const char MetadataQuery::kZoneAttribute[] =
+    "/computeMetadata/v1/instance/zone";
+constexpr const char MetadataQuery::kClusterNameAttribute[] =
+    "/computeMetadata/v1/instance/attributes/cluster-name";
+constexpr const char MetadataQuery::kRegionAttribute[] =
+    "/computeMetadata/v1/instance/region";
+constexpr const char MetadataQuery::kInstanceIdAttribute[] =
+    "/computeMetadata/v1/instance/id";
+constexpr const char MetadataQuery::kIPv6Attribute[] =
+    "/computeMetadata/v1/instance/network-interfaces/0/ipv6s";
+
+MetadataQuery::MetadataQuery(
+    std::string attribute, grpc_polling_entity* pollent,
+    absl::AnyInvocable<void(std::string /* attribute */,
+                            std::string /* result */)>
+        callback,
+    Duration timeout)
+    : InternallyRefCounted<MetadataQuery>(nullptr, 2),
+      attribute_(std::move(attribute)),
+      callback_(std::move(callback)) {
+  GRPC_CLOSURE_INIT(&on_done_, OnDone, this, nullptr);
+  auto uri = URI::Create("http", "metadata.google.internal.", attribute_,
+                         {} /* query params */, "" /* fragment */);
+  GPR_ASSERT(uri.ok());  // params are hardcoded
+  grpc_http_request request;
+  memset(&request, 0, sizeof(grpc_http_request));
+  grpc_http_header header = {const_cast<char*>("Metadata-Flavor"),
+                             const_cast<char*>("Google")};
+  request.hdr_count = 1;
+  request.hdrs = &header;
+  // The http call is local. If it takes more than one sec, it is probably not
+  // on GCP.
+  http_request_ = HttpRequest::Get(
+      std::move(*uri), nullptr /* channel args */, pollent, &request,
+      Timestamp::Now() + timeout, &on_done_, &response_,
+      RefCountedPtr<grpc_channel_credentials>(
+          grpc_insecure_credentials_create()));
+  http_request_->Start();
+}
+
+MetadataQuery::~MetadataQuery() { grpc_http_response_destroy(&response_); }
+
+void MetadataQuery::Orphan() {
+  http_request_.reset();
+  Unref();
+}
+
+void MetadataQuery::OnDone(void* arg, absl::Status error) {
+  auto* self = static_cast<MetadataQuery*>(arg);
+  std::string result;
+  if (!error.ok()) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_metadata_query_trace)) {
+      gpr_log(GPR_INFO, "MetadataServer Query failed for %s: %s",
+              self->attribute_.c_str(), StatusToString(error).c_str());
+    }
+  } else if (self->response_.status != 200) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_metadata_query_trace)) {
```

Comment:
nit: as long as this tracer flag is enabled, it will probably be helpful to log all MD server query results, and the HTTP status will probably be useful to include in the log too.

e.g. suggest replacing:

```
if (!error.ok()) {
  ...
} else if (self->response_.status != 200) {
  ...
}
```

with just one log at the top:

```
if (GRPC_TRACE_FLAG_ENABLED(grpc_metadata_query_trace)) {
    gpr_log(GPR_INFO, "MetadataServer Query for %s: http status: %d, error: %s",
            self->attribute_.c_str(), self->response_.status, StatusToString(error).c_str());
}
```

Reply:
done